clang/lib/Sema/SemaAccess.cpp

```diff
@@ -776,6 +777,8 @@ static AccessResult HasAccess(Sema &S,
     // [B3] and [M3]
     } else {
       assert(Access == AS_protected);
+      if (!ECRecord->getDefinition())
```

Comment:
Should these changes be made in community clang? This doesn't seem SYCL specific

Reply:
I have tried but I there is no upstream class attribute that accepts an arbitrary expression. Couldn't find a test case that crashes. The closest one I could find is https://godbolt.org/z/qxEjjvf73 , but it doesn't crash. It returns the expected error.

Reply:
> When is ECRecord not defined here?

This function is checking if the entity is accessible with the specific `AccessSpecifier`.  If it's `protected` then we need to make sure that the type is complete. A type is incomplete if `!getDefinition()` .

 > When is `ECRecord `not defined here? And why is this restriction only required for `protected` access?

Please see comments at line 858 concerning [M3] and [B3].
`IsDerivedFromInclusive` checks whether a class is derived from another class. The question is, should `IsDerivedFromInclusive` have a precondition on `Derived` being a complete type. I think it should.

Reply:
See https://eel.is/c++draft/class.access [M3] and https://eel.is/c++draft/class.access#class.protected [B3]